httpx/middleware/digest_auth.py

```diff
@@ -0,0 +1,188 @@
+import hashlib
+import os
+import re
+import time
+import typing
+from urllib.request import parse_http_list
+
+from ..exceptions import ProtocolError
+from ..models import AsyncRequest, AsyncResponse, StatusCode
+from ..utils import to_bytes, to_str, unquote
+from .base import BaseMiddleware
+
+
+class DigestAuth(BaseMiddleware):
+    ALGORITHM_TO_HASH_FUNCTION: typing.Dict[str, typing.Callable] = {
+        "MD5": hashlib.md5,
+        "MD5-SESS": hashlib.md5,
+        "SHA": hashlib.sha1,
+        "SHA-SESS": hashlib.sha1,
+        "SHA-256": hashlib.sha256,
+        "SHA-256-SESS": hashlib.sha256,
+        "SHA-512": hashlib.sha512,
+        "SHA-512-SESS": hashlib.sha512,
+    }
+
+    def __init__(
+        self, username: typing.Union[str, bytes], password: typing.Union[str, bytes]
+    ) -> None:
+        self.username = to_bytes(username)
+        self.password = to_bytes(password)
+        self._num_401_responses = 0
```

Comment:
Okay, so we can't have *mutable* state on our middleware classes.

I assume the usage of `self._num_401_responses` here is why the async case is currently failing (and multi-threaded likely would do too.)

I would probably start by removing this property entirely, and removing the associated branching in `__call__` (If the very first response we get back is OK, then return it, otherwise build a request with the auth header, and return whatever response we get back from that.)

Once that's in line, I'll give this another review.

Great stuff. 👍

Reply:
Inevitably people will want to store mutable state on middleware, should I open a design discussion issue for us to nail those details and document them for others as well?

Reply:
I'd say go for it, even if the stance is "No mutable state, just act on it and pass it through"